### DISCUSSION TITLE
Jinja templating for messages in PubSubPublishOperator

### DISCUSSION THREAD

**Discussion by CyberHippo** [apache/airflow#22303] [Q&A] [answered]
Hi,

I would like to use Jinja templating for the messages sent by the `PubSubPublishOperator` in my DAG.

I use the following DAG as a minimal example:

```py
from airflow import DAG
from datetime import datetime
from airflow.contrib.operators.kubernetes_pod_operator import KubernetesPodOperator
from airflow.contrib.operators.pubsub_operator import PubSubPublishOperator
from base64 import b64encode as b64e
import json

with DAG(
    dag_id = 'pubsub-message-jinja',
    schedule_interval = "* * * * *",
    default_args = {
        'owner': 'admin',
        'start_date': datetime.strptime('2022-01-01', '%Y-%m-%d'),
    },
    catchup = False,
) as dag:

    message_raw = { "data": "{{ ds }}" }
    message_encoded = b64e(json.dumps(message_raw).encode("utf-8"))

    pubsub_publisher = PubSubPublishOperator(
        gcp_conn_id="dummy_gcp_conn_id",
        task_id="pubsub-notification",
        project="dummy_gcp_project_id",
        topic="dummy_gcp_topic_name",
        messages=[
          {'data': message_encoded },
        ]
    )

    pubsub_publisher
```

The message is successfully sent to the topic. However the Jinja templating did not occur and the message received is `{"data": "{{ ds }}"}`.

According to the [PubSubPublishOperator docs](https://airflow.apache.org/docs/apache-airflow-providers-google/stable/_api/airflow/providers/google/cloud/operators/pubsub/index.html#airflow.providers.google.cloud.operators.pubsub.PubSubPublishMessageOperator), the field `message` is templated and I would expect `ds` to be evaluated before the message is sent to PubSub.

Is there something I am missing ?

Regards,
Hippolyte V.

**Comment by josh-fell** [comment, marked as answer]
The issue that you are executing top-level code on a templated string so ` b64e(json.dumps(message_raw).encode("utf-8"))` evaluates the literal string "{{ ds }}" rather than the rendered value that you are expecting. Templated strings are not rendered/evaluated until the task begins to execute. 

If you need enforce the encoding on the messages, you can try defining a `user_defined_macros` or `user_defined_filters` to use with the Jinja templating.

>user_defined_macros ([dict](https://docs.python.org/3/library/stdtypes.html#dict)) – a dictionary of macros that will be exposed in your jinja templates. For example, passing dict(foo='bar') to this argument allows you to {{ foo }} in all jinja templates related to this DAG. Note that you can pass any type of object here.

>user_defined_filters ([dict](https://docs.python.org/3/library/stdtypes.html#dict)) – a dictionary of filters that will be exposed in your jinja templates. For example, passing dict(hello=lambda name: 'Hello %s' % name) to this argument allows you to {{ 'world' | hello }} in all jinja templates related to this DAG.

Another user wrote their own `user_defined_filter` for the encoding and seemed to work. [This is the discussion](https://github.com/apache/airflow/discussions/21194) if you find it helpful.

**Comment by CyberHippo** [comment]
@josh-fell Thank you for your helpful answer! I got the example working with a `user_defined_filter` and ended up using a `user_defined_macros` for my use case.

**Reply by alexwu0911** [reply]
@CyberHippo  Can you show your modified code ?

**Reply by CyberHippo** [reply]
@alexwu0911 Sorry for the delay. Please find my code below:
```python
from airflow import DAG
from datetime import datetime
from airflow.contrib.operators.kubernetes_pod_operator import KubernetesPodOperator
from airflow.providers.google.cloud.operators.pubsub import PubSubPublishMessageOperator
import json


def message_encode(ds):
    message = { "foo": str(ds) }
    return json.dumps(message).encode('utf-8')


with DAG(
    dag_id="pubsub-message-jinja",
    schedule_interval="* * * * *",
    default_args={
        "owner": "admin",
        "start_date": datetime.strptime("2022-01-01", "%Y-%m-%d"),
    },
    catchup=False,
    user_defined_macros={"pubsub_encode": message_encode},
    render_template_as_native_obj=True,
) as dag:

    message = "{{ pubsub_encode(ds) }}"

    pubsub_publisher = PubSubPublishMessageOperator(
        gcp_conn_id="dummy_gcp_conn_id",
        task_id="pubsub-notification",
        project_id="dummy_gcp_project_id",
        topic="dummy_gcp_topic_name",
        messages=[
            {"data": message},
        ],
    )

    pubsub_publisher
```

**Reply by atorres-eqrx** [reply]
Thanks for sharing your solution @CyberHippo. I have a slightly different use cases and I'm wondering if this solution is extensible to handling dictionaries. 

E.g.

```python
def message_encode(message: dict):
    return json.dumps(message).encode('utf-8')


with DAG(
    dag_id="pubsub-message-jinja",
    schedule_interval="* * * * *",
    default_args={
        "owner": "admin",
        "start_date": datetime.strptime("2022-01-01", "%Y-%m-%d"),
    },
    catchup=False,
    user_defined_macros={"pubsub_encode": message_encode},
    render_template_as_native_obj=True,
) as dag:

    payload = {
      "key":"value"
    }
    
    message = "{{ pubsub_encode(payload) }}"

    pubsub_publisher = PubSubPublishMessageOperator(
        gcp_conn_id="dummy_gcp_conn_id",
        task_id="pubsub-notification",
        project_id="dummy_gcp_project_id",
        topic="dummy_gcp_topic_name",
        messages=[
            {"data": message},
        ],
    )

    pubsub_publisher
```

**Reply by CyberHippo** [reply]
@atorres-eqrx I confirm that you can handle dictionaries. We are doing so by including the logic in the macro.

**Comment by hankehly** [comment]
I thought allowing string-type message data would make the operator easier to use, [so I submitted a PR.](https://github.com/apache/airflow/pull/30981)

**Comment by shekarraj3** [comment]
Below is the my code i am unable to publish message in composer 2.0
can anyone help how to resolve 

from airflow import DAG
from datetime import datetime, timedelta
from airflow.models import Variable
from airflow import configuration as conf
from airflow.operators.bash import BashOperator
from airflow.operators.subdag import SubDagOperator
import logging
import subprocess
import os
from random import randint
#from airflow.hooks import PostgresHook
from airflow.providers.postgres.hooks.postgres import PostgresHook
from airflow.utils.trigger_rule import TriggerRule
#from airflow.contrib.operators.dataproc_operator import DataprocClusterCreateOperator, DataprocClusterDeleteOperator
from airflow.providers.google.cloud.operators.dataproc import DataprocDeleteClusterOperator
#from airflow.operators import PythonOperator,  BashOperator, PostgresOperator, SubDagOperator
from airflow.operators.python import PythonOperator
from airflow.operators.bash import BashOperator
from airflow.providers.postgres.operators.postgres import PostgresOperator

#from airflow.contrib.operators.dataproc_operator import DataProcSparkOperator
from airflow.models import DagRun
from airflow.executors.celery_executor import CeleryExecutor
from airflow.operators.python import BranchPythonOperator
#from airflow.operators.dummy_operator import EmptyOperator
from airflow.operators.empty import EmptyOperator
import ast
#from airflow.contrib.operators.pubsub_operator import PubSubPublishOperator
from airflow.providers.google.cloud.operators.pubsub import PubSubPublishMessageOperator
from airflow.providers.google.cloud.hooks.pubsub import PubSubHook
from base64 import b64encode
from base64 import b64encode as b64e
import base64
import json
#from operators import MyDataProcSparkOperator
#from util_version2 import *
import time


default_args = {
        'owner': 'cloud-composer',
        'depends_on_past': True,
        #'start_date': datetime.now(),
        'start_date': datetime(2018, 6, 1),
        'provide_context': True,
        'retries': DAG_RETRIES,
        'email': [DAG_EMAIL_ID],
        'email_on_failure': True,
        'email_on_retry': False,
        'params': {"practice": "value2"},
        'retry_delay': timedelta(seconds=int(DAG_RETRY_DELAY)),
        'execution_timeout': timedelta(minutes=EXECUTION_TASK_TIMEOUT)
}

DAG_ID='%s_PUBSUB_TEST'%REGISTRY


dag = DAG(
	DAG_ID,
	default_args=default_args,
	start_date=datetime(2018, 6, 1),
	schedule_interval=None)


#success_pubsub_msg=[{"data":"{{ task_instance.xcom_pull('get_pubsub_success_mesg', key='message')}}"}]
#success_pubsub_msg={"data":"{{ task_instance.xcom_pull('get_pubsub_success_mesg', key='message')}}"}
success_pubsub_msg={"data":"{{ task_instance.xcom_pull('get_pubsub_success_mesg', key='return_value')}}".encode('utf-8')}
def getPubSubSuccessMessage(**context):
	practice_args = dict()
	practice_args['practice_list'] = "111"
	practice_args['message_id'] = "6804860098820738"
	practice_args['module_name'] = "module_name"
	practice_args['src_queue'] = "topic"
	practice_args['run_id'] = "temp_runid"
	practice_args['status'] = "SUCCESS"
	practice_args_json = json.dumps(practice_args)
	print('practice_args_json',practice_args_json)
	print(json.dumps(practice_args).encode('utf-8'))
	#context['ti'].xcom_push(key="message", value=bytes(json.dumps(practice_args).encode('utf-8')))
	#context['ti'].xcom_push(key="message", value=json.dumps(practice_args).encode('utf-8'))
	encoded_string = base64.b64encode(practice_args_json.encode('utf-8')).decode()
	#encoded_string = practice_args_json.encode('utf-8')
	return encoded_string

get_pubsub_success_mesg = PythonOperator(
                        task_id='get_pubsub_success_mesg', python_callable=getPubSubSuccessMessage, trigger_rule=NONE_FAILED_TRIGGER_RULE,dag=dag)

PubSubPublishSuccess = PubSubPublishMessageOperator(task_id='publish-messages_success',
	topic=PUB_SUB_TOPIC,
	project_id=PROJECT_ID,
	#messages=success_pubsub_msg,
	#messages=[{"data": success_pubsub_msg.encode('utf-8')},],
	messages=[{"data": success_pubsub_msg},],
	dag=dag)

get_pubsub_success_mesg >> PubSubPublishSuccess
